src/hotspot/share/gc/g1/g1HeapRegionRemSet.hpp

```diff
@@ -47,7 +47,8 @@ class G1HeapRegionRemSet : public CHeapObj<mtGC> {
   G1CardSetMemoryManager _card_set_mm;
 
   // The set of cards in the Java heap
-  G1CardSet _card_set;
+  G1CardSet* _card_set;
+  G1CardSet* _saved_card_set;
```

Comment:
I think the reason for have *two* fields instead of one is that a region can go through old<->young transitions multiple times. A young-region uses the second field to save the card-set used when the same region become old. I wonder if this "detail" can be extracted out of remset, sth like an array of card-set pointers maintained outside of remset. IOW, it seems an abstraction leakage if remset knows there are two "kinds" of card-set.